scipy/interpolate/_bspl.pyx

```diff
@@ -0,0 +1,410 @@
+"""
+Routines for evaluating and manipulating B-splines.
+
+"""
+
+import numpy as np
+cimport numpy as cnp
+
+cimport cython
+
+cdef extern from "src/__fitpack.h":
+    void _deBoor_D(double *t, double x, int k, int ell, int m, double *result)
+
+cdef double nan = np.nan
+
+ctypedef double complex double_complex
+
+ctypedef fused double_or_complex:
+    double
+    double complex
+
+
+#------------------------------------------------------------------------------
+# B-splines
+#------------------------------------------------------------------------------
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline int find_interval(double[::1] t,
+                       int k,
+                       double xval,
+                       int prev_l,
+                       int extrapolate) nogil:
+    """
+    Find an interval such that t[interval] <= xval < t[interval+1].
+
+    Uses a linear search with locality, see fitpack's splev.
+
+    Parameters
+    ----------
+    t : ndarray, shape (nt,)
+        Knots
+    k : int
+        B-spline degree
+    xval : double
+        value to find the inteval for
+    prev_l : int
+        interval where the previous value was located.
+        if unknown, use any value < k to start the search.
+    extrapolate : int
+        whether to return the last or the first interval if xval
+        is out of bounds.
+
+    Returns
+    -------
+    interval : int
+        Suitable interval or -1 if xval was nan.
+
+    """
+    cdef:
+        int l
+        int n = t.shape[0] - k - 1
+        double tb = t[k]
+        double te = t[n]
+
+    if xval != xval:
+        # nan
+        return -1
+
+    if (xval < tb) or (xval > te):
+        if extrapolate:
+            pass
+        else:
+            return -1
+
+    l = prev_l if k < prev_l < n else k
+
+    # xval is in support, search for interval s.t. t[interval] <= xval < t[l+1]
+    while(xval < t[l] and l != k):
+        l -= 1
+
+    l += 1
+    while(xval >= t[l] and l != n):
+        l += 1
+
+    return l-1
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+def evaluate_spline(double[::1] t,
+             double_or_complex[:, ::1] c,
+             int k,
+             double[::1] xp,
+             int der,
+             int extrapolate,
+             double_or_complex[:, ::1] out):
+    """
+    Evaluate a spline in the B-spline basis.
+
+    Parameters
+    ----------
+    t : ndarray, shape (n+k+1)
+        knots
+    c : ndarray, shape (n, m)
+        B-spline coefficients
+    xp : ndarray, shape (s,)
+        Points to evaluate the spline at.
+    der : int
```

Comment:
Renamed it `nu` to agree with *Poly and their subclasses (monotone interpolators).